libs/agno/agno/vectordb/surrealdb/surrealdb.py

```diff
@@ -0,0 +1,392 @@
+from collections.abc import AsyncGenerator, Generator
+from contextlib import asynccontextmanager, contextmanager
+from typing import Any, Dict, Final, List, Optional, Union
+
+from surrealdb import (
+    AsyncHttpSurrealConnection,
+    AsyncSurreal,
+    AsyncWsSurrealConnection,
+    BlockingHttpSurrealConnection,
+    BlockingWsSurrealConnection,
+    Surreal,
+)
+
+from agno.document import Document
+from agno.embedder import Embedder
+from agno.utils.log import log_debug, log_error, log_info
+from agno.vectordb.base import VectorDb
+from agno.vectordb.distance import Distance
+
+
+class SurrealDb(VectorDb):
+    """SurrealDB Vector Database implementation supporting both sync and async operations"""
+
+    # SQL Query Constants
+    CREATE_TABLE_QUERY: Final[str] = """
+        DEFINE TABLE IF NOT EXISTS {collection} SCHEMAFUL;
+        DEFINE FIELD IF NOT EXISTS content ON {collection} TYPE string;
+        DEFINE FIELD IF NOT EXISTS embedding ON {collection} TYPE array<float>;
+        DEFINE FIELD IF NOT EXISTS meta_data ON {collection} TYPE object;
+        DEFINE INDEX IF NOT EXISTS vector_idx ON {collection} FIELDS embedding HNSW DIMENSION {dimensions} DIST COSINE;
+    """
+
+    DOC_EXISTS_QUERY: Final[str] = """
+        SELECT * FROM {collection}
+        WHERE content = $content
+        LIMIT 1
+    """
+
+    NAME_EXISTS_QUERY: Final[str] = """
+        SELECT * FROM {collection}
+        WHERE meta_data.name = $name
+        LIMIT 1
+    """
+
+    ID_EXISTS_QUERY: Final[str] = """
+        SELECT * FROM {collection}
+        WHERE id = $id
+        LIMIT 1
+    """
+
+    UPSERT_QUERY: Final[str] = """
+        UPSERT {thing}
+        SET content = $content,
+            embedding = $embedding,
+            meta_data = $meta_data
+    """
+
+    SEARCH_QUERY: Final[str] = """
+        SELECT
+            content,
+            meta_data,
+            vector::distance::knn() as distance
+        FROM {collection}
+        WHERE embedding <|{limit},{distance}|> $query_embedding
```

Comment:
Specifying here a `distance` function makes the query use brute force, instead of the HNSW index. The distance function should be set in the index instead.

```suggestion
        WHERE embedding <|{limit}|> $query_embedding
```

Reply:
Thanks a lot for the clarifications.